src/states/infinitepartitionfunction.jl

```diff
@@ -0,0 +1,240 @@
+) where {S<:Union{ElementarySpace,Int}}
+    return InfinitePartitionFunction(
+        randn, ComplexF64, fill(Nspace, unitcell), fill(Espace, unitcell)
+    )
+end
+function InfinitePartitionFunction(
+    f, T, Nspace::S, Espace::S=Nspace; unitcell::Tuple{Int,Int}=(1, 1)
+) where {S<:Union{ElementarySpace,Int}}
+    return InfinitePartitionFunction(f, T, fill(Nspace, unitcell), fill(Espace, unitcell))
+end
+
+## Shape and size
+Base.size(T::InfinitePartitionFunction) = size(T.A)
+Base.size(T::InfinitePartitionFunction, i) = size(T.A, i)
+Base.length(T::InfinitePartitionFunction) = length(T.A)
+Base.eltype(T::InfinitePartitionFunction) = eltype(typeof(T))
+Base.eltype(::Type{<:InfinitePartitionFunction{T}}) where {T} = T
+function VectorInterface.scalartype(::Type{T}) where {T<:InfinitePartitionFunction}
+    return scalartype(eltype(T))
+end
+
+## Copy
+Base.copy(T::InfinitePartitionFunction) = InfinitePartitionFunction(copy(T.A))
+function Base.similar(T::InfinitePartitionFunction, args...)
+    return InfinitePartitionFunction(similar(T.A, args...))
+end
+function Base.repeat(T::InfinitePartitionFunction, counts...)
+    return InfinitePartitionFunction(repeat(T.A, counts...))
+end
+
+Base.getindex(T::InfinitePartitionFunction, args...) = Base.getindex(T.A, args...)
+Base.setindex!(T::InfinitePartitionFunction, args...) = (Base.setindex!(T.A, args...); T)
+Base.axes(T::InfinitePartitionFunction, args...) = axes(T.A, args...)
+function eachcoordinate(x::InfinitePartitionFunction)
+    return collect(Iterators.product(axes(x)...))
+end
+function eachcoordinate(x::InfinitePartitionFunction, dirs)
+    return collect(Iterators.product(dirs, axes(x, 1), axes(x, 2)))
+end
+TensorKit.space(t::InfinitePartitionFunction, i, j) = space(t[i, j], 1)
+
+## Math
+function Base.:+(ψ₁::InfinitePartitionFunction, ψ₂::InfinitePartitionFunction)
+    return InfinitePartitionFunction(ψ₁.A + ψ₂.A)
+end
+function Base.:-(ψ₁::InfinitePartitionFunction, ψ₂::InfinitePartitionFunction)
+    return InfinitePartitionFunction(ψ₁.A - ψ₂.A)
+end
+Base.:*(α::Number, ψ::InfinitePartitionFunction) = InfinitePartitionFunction(α * ψ.A)
+Base.:/(ψ::InfinitePartitionFunction, α::Number) = InfinitePartitionFunction(ψ.A / α)
+function LinearAlgebra.dot(ψ₁::InfinitePartitionFunction, ψ₂::InfinitePartitionFunction)
+    return dot(ψ₁.A, ψ₂.A)
+end
+LinearAlgebra.norm(ψ::InfinitePartitionFunction) = norm(ψ.A)
+
+## (Approximate) equality
+function Base.:(==)(ψ₁::InfinitePartitionFunction, ψ₂::InfinitePartitionFunction)
+    return all(zip(ψ₁.A, ψ₂.A)) do (p₁, p₂)
+        return p₁ == p₂
+    end
+end
+function Base.isapprox(
+    ψ₁::InfinitePartitionFunction, ψ₂::InfinitePartitionFunction; kwargs...
+)
+    return all(zip(ψ₁.A, ψ₂.A)) do (p₁, p₂)
+        return isapprox(p₁, p₂; kwargs...)
+    end
+end
+
+# Used in _scale during OptimKit.optimize
+function LinearAlgebra.rmul!(ψ::InfinitePartitionFunction, α::Number)
+    rmul!(ψ.A, α)
```

Comment:
```suggestion
    rmul!.(ψ.A, α)
```
Also inplace usage of the tensors